### ISSUE TITLE
global.Proxy is an MOBoxedObject of clas, not the ES6 Proxy object

### ISSUE THREAD

**Issue by cameron** [logancollins/Mocha#36]
am i misunderstanding the claims made here? https://developer.sketchapp.com/guides/cocoascript/

or is my obj-c environment doing something shady?

would love to have a proper Proxy in cocoascript!

**Comment by mathieudutour** [comment]
Hum weird, I'm pretty sure I have a special case for it (same as `Function` and `Object`): https://github.com/BohemianCoding/CocoaScript/blob/develop/src/framework/mocha/MochaRuntime.m#L1038

**Comment by matt-curtis** [comment]
> Really? I've never really seen Proxy used in the wild yet

I both tried to write some code that uses it and also tried to pull in a library that used it, and they failed due to this bug.

Also the line you linked to doesn't fix the problem because for whatever reason, `NSClassFromString("Proxy")` returns `nil`, so the if block is never entered. Personally, I think it's bad for future JS compatibility that any native method or property on the global object can be overridden by an Objective-C class, C function, or other type.

Ideally Objective-C symbols would have been namespaced by default and never able to override default global object properties at all, but since that was an early implementation choice and too many scripts depend on them being global now perhaps the best solution is for the global object to continue to work as it does now, but also provide a proxy object as a property on the global object (i.e. `globalObject.unmodified`) that will always return the unmodified global object, free of any dynamic look up of symbols.

**Comment by mathieudutour** [comment]
yeah you are rightI was thinking about taking all the built-in JS objects and making a special case for them.

Right now the code to find a global is doing the following:
- check if the name is `__mocha__` -> default to the JS global object
- check if mocha has an object with the name attached to it -> return it
- check if a obj-c class exists and if the classname is not `Object` or `Function` -> return the class
- check if a c function or a C constant or a enum exists -> return it (I'm guessing that's what happens with `Proxy` since `_NCSFNumber` is a C constant)
- default to the JS global object

I think it should be:
- check if the name is `__mocha__` or any built-in JS object -> default to the JS global object
- check if mocha has an object with the name attached to it -> return it
- check if a obj-c class exists -> return the class
- check if a c function or a C constant or a enum exists -> return it (I'm guessing that's what happens with `Proxy` since `_NCSFNumber` is a C constant)
- default to the JS global object

That would be a breaking change but I don't believe it will actually break anything

**Comment by matt-curtis** [comment]
Your suggestion is what I had in mind at first, however it still has the problem in that the reverse can happen: default global object properties can then override Objective-C symbols, with no way to access the desired Objective-C symbol. Whatever solution is chosen, though, needs to be as forward-compatible with changes to the ECMAScript global object/built-in objects and any changes that might occur in the default Foundation/etc. frameworks.

One way to buffer the impact of a breaking change is to finally put a use to the ‘bundleVersion’ key in the manifest.

**Comment by mathieudutour** [comment]
I don't think the bundleVersion comes into play since it's about compatibility with macOS (both the obj-c API and the JavaScriptCore API). The way the bundle is layed out doesn't matter.

> default global object properties can then override Objective-C symbols

The nice thing about the Obj-C API is that everything you want to access is prefixed with `NS` or `MS` are whatever. There are really few symbols that aren't and when they are not, it's really low level symbols that you don't want to use directly anyway.

This change only impact plugin devs that write in JS and I'm convinced it's more confusing not to be able to access a JS global object than a obj-c symbol.

Moreover, the worst bug that can happen is when there is a new macOS update that include a new symbol that override a JS object. That happened with `Function` which was added in macOS 10.13 IIRC. On the other hand, if a new JS Object is added, I think it's fine to wait for a Sketch update, it won't break existing plugins.

I think I just convinced myself that it's the right thing to do :)

**Comment by cameron** [comment]
ty, folks!

**Comment by matt-curtis** [comment]
> The nice thing about the Obj-C API is that everything you want to access is prefixed with NS or MS are whatever. There are really few symbols that aren't and when they are not, it's really low level symbols that you don't want to use directly anyway.

I agree, for the most part... Seems like a reasonable trade-off.
